internal/memhttp/option.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2021-2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memhttp
+
+import (
+	"log"
+	"time"
+)
+
+type config struct {
+	DisableTLS     bool
+	DisableHTTP2   bool
+	CleanupTimeout time.Duration
+	ErrorLog       *log.Logger
+}
+
+// An Option configures a Server.
+type Option interface {
+	apply(*config)
+}
+
+type optionFunc func(*config)
+
+func (f optionFunc) apply(cfg *config) { f(cfg) }
+
+// WithoutHTTP2 disables HTTP/2 on the server and client.
+func WithoutHTTP2() Option {
+	return optionFunc(func(cfg *config) {
+		cfg.DisableHTTP2 = true
+	})
+}
+
+// WithOptions composes multiple Options into one.
+func WithOptions(opts ...Option) Option {
+	return optionFunc(func(cfg *config) {
+		for _, opt := range opts {
+			opt.apply(cfg)
+		}
+	})
+}
+
+// WithCleanupTimeout customizes the default five-second timeout for the
+// server's Cleanup method. It's most useful with the memhttptest subpackage.
+func WithCleanupTimeout(d time.Duration) Option {
```

Comment:
If we move `config` into an internal package (accessible only to `memhttp` and `memhttptest` packages), then we could move this option into `memhttptest`, so that you can only use it if you import `memhttptest`, which I think makes the final remark in the Go docs unnecessary.

Further, we could make it so that it is only applied in `memhttptest` by having this value actually used in the `t.Cleanup` function, instead of in `svr.Shutdown`.

I think that leaves the API in `memhttp` a little more "pure", and the testing helpers are all isolated to the `memhttptest` package.

Reply:
I've renamed the option to `WithShutdownTimeout`. It's used to set a timer on the context so simplifies users calling shutdown and guarantees the context get's cancelled. I think this is valuable normal server usage, removed the test desc.

Reply:
>  I think this is valuable normal server usage, removed the test desc.

I don't think that's the right call. For one, it is oddly inconsistent with `http.Server.Shutdown`: it's weird that we _change_ the context that the user supplies. Also, it puts the timeout that the method waits far from the callsite, which can impede readability. If the user wants a shutdown timeout, they add a timeout to the `ctx` right at the callsite.

The only value for it is in tests, where `Shutdown` is called transparently as part of test cleanup. In this case, the user doesn't have control over the callsite, so the duration to wait must be provided up-front when the cleanup action is configured.

Reply:
It doesn't modify the original context, just guarantees it always cancels. It's bounded to the limit but can always be reduced by the caller if desired. I thought it was a neat simplification but will revert.

Reply:
Wasn't sure how to restrict Cleanup only to the test server without more changes. So exposed Cleanup method on the main server.

Reply:
Thank you for making those changes.

Restricting `Cleanup` only to the test server can be easily done by _removing_ it (and the associated field from `memhttp.Server`) and make the timeout _only_ used in the call to `t.Cleanup` in the test package.

The cleanup timeout is really only needed for that one place. If a test wants to explicitly stop the server before that cleanup happens, it can use the server's `Shutdown` and `Close` methods.

Reply:
How do you want the test server to get the `CleanupTimeout` on the `*internal.Config`? Options are currently built within `NewServer`.

Reply:
Let's punt on it. I can put up a PR later to show what I mean and you and Akshay can see what you think at that time.